File: Tutorial_Sample_Code/ViewerObjectModel/ParametersSubrpt/CS/CS_Win_CRVObjMod_ParametersSubrpt/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using CrystalDecisions.CrystalReports.Engine;
using CrystalDecisions.Shared;
using System.Collections;

namespace CS_Win_CRVObjMod_ParametersSubrpt
{
    public partial class Form1 : Form
    {
        private const string PARAMETER_FIELD_NAME = "City";
        private const string SUBREPORT_PARAMETER_FIELD_NAME = "OrderDateRange";
        private const string SUBREPORT_NAME = "CustomerOrders";


        public Form1()
        {
            InitializeComponent();
        }

        private void SetCurrentValuesForParameterField(ParameterFields parameterFields, ArrayList arrayList)
        {
            ParameterValues currentParameterValues = new ParameterValues();
            foreach (object submittedValue in arrayList)
            {
                ParameterDiscreteValue parameterDiscreteValue = new ParameterDiscreteValue();
                parameterDiscreteValue.Value = submittedValue.ToString();
                currentParameterValues.Add(parameterDiscreteValue);
            }
            ParameterField parameterField = parameterFields[PARAMETER_FIELD_NAME];
            parameterField.CurrentValues = currentParameterValues;

        }


        private void ConfigureCrystalReports()
        {
            ArrayList arrayList = new ArrayList();
            arrayList.Add("Paris");
            arrayList.Add("Tokyo");

            string startDate = "8/1/1997";
            string endDate = "8/31/1997";

            string reportPath = Application.StartupPath + "\\" + "CustomersByCity.rpt";
            crystalReportViewer.ReportSource = reportPath;
            ParameterFields parameterFields = crystalReportViewer.ParameterFieldInfo;
            SetCurrentValuesForParameterField(parameterFields, arrayList);
            SetDateRangeForOrders(parameterFields, startDate, endDate);
            defaultParameterValuesList.DataSource = GetDefaultValuesFromParameterField(parameterFields);

        }

        private void Form1_Load(object sender, EventArgs e)
        {
            ConfigureCrystalReports();
        }

        private void crystalReportViewer_Load(object sender, EventArgs e)
        {

        }

        private ArrayList GetDefaultValuesFromParameterField(ParameterFields parameterFields)
        {
            ParameterField parameterField = parameterFields[PARAMETER_FIELD_NAME];
            ParameterValues defaultParameterValues = parameterField.DefaultValues;
            ArrayList arrayList = new ArrayList();
            foreach (ParameterValue parameterValue in defaultParameterValues)
            {
                if (!parameterValue.IsRange)
                {
                    ParameterDiscreteValue parameterDiscreteValue = (ParameterDiscreteValue)parameterValue;
                    arrayList.Add(parameterDiscreteValue.Value.ToString());
                }
            }
            return arrayList;
        }

        private void redisplay_Click(object sender, EventArgs e)
        {
            ArrayList arrayList = new ArrayList();
            foreach (string item in defaultParameterValuesList.SelectedItems)
            {
                arrayList.Add(item);
            }

            string startDate = orderStartDate.Text;
            string endDate = orderEndDate.Text;

            crystalReportViewer.ReportSource = Application.StartupPath + "\\" + "CustomersByCity.rpt";
            ParameterFields parameterFields = crystalReportViewer.ParameterFieldInfo;
            SetCurrentValuesForParameterField(parameterFields, arrayList);

            SetDateRangeForOrders(parameterFields, startDate, endDate);

        }

        private void SetDateRangeForOrders(ParameterFields parameterFields, string startDate, string endDate)
        {
            ParameterRangeValue parameterRangeValue = new ParameterRangeValue();
            parameterRangeValue.StartValue = startDate;
            parameterRangeValue.EndValue = endDate;
            parameterRangeValue.LowerBoundType = RangeBoundType.BoundInclusive;
            parameterRangeValue.UpperBoundType = RangeBoundType.BoundInclusive;
            ParameterField parameterField = parameterFields[SUBREPORT_PARAMETER_FIELD_NAME, SUBREPORT_NAME];
            parameterField.CurrentValues.Clear();
            parameterField.CurrentValues.Add(parameterRangeValue);

        }

    }
}

File: Tutorial_Sample_Code/ViewerObjectModel/ParametersSubrpt/VB/VB_Win_CRVObjMod_ParametersSubrpt/Form1.vb
Imports CrystalDecisions.CrystalReports.Engine
Imports CrystalDecisions.Shared
Imports System.Collections


Public Class Form1

    Private Const PARAMETER_FIELD_NAME As String = "City"
    Private Const SUBREPORT_PARAMETER_FIELD_NAME As String = "OrderDateRange"
    Private Const SUBREPORT_NAME As String = "CustomerOrders"


    Private Sub ConfigureCrystalReports()
        Dim myArrayList As ArrayList = New ArrayList()
        myArrayList.Add("Paris")
        myArrayList.Add("Tokyo")

        Dim startDate As String = "8/1/1997"
        Dim endDate As String = "8/31/1997"

        Dim reportPath As String = Application.StartupPath & "\" & "CustomersByCity.rpt"
        myCrystalReportViewer.ReportSource = reportPath
        Dim myParameterFields As ParameterFields = myCrystalReportViewer.ParameterFieldInfo
        SetCurrentValuesForParameterField(myParameterFields, myArrayList)
        SetDateRangeForOrders(myParameterFields, startDate, endDate)
        defaultParameterValuesList.DataSource = GetDefaultValuesFromParameterField(myParameterFields)

    End Sub

    Private Function GetDefaultValuesFromParameterField(ByVal myParameterFields As ParameterFields) As ArrayList
        Dim myParameterField As ParameterField = myParameterFields(PARAMETER_FIELD_NAME)
        Dim defaultParameterValues As ParameterValues = myParameterField.DefaultValues
        Dim myArrayList As ArrayList = New ArrayList()
        For Each myParameterValue As ParameterValue In defaultParameterValues
            If (Not myParameterValue.IsRange) Then
                Dim myParameterDiscreteValue As ParameterDiscreteValue = CType(myParameterValue, ParameterDiscreteValue)
                myArrayList.Add(myParameterDiscreteValue.Value.ToString())
            End If
        Next
        Return myArrayList
    End Function

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        ConfigureCrystalReports()
    End Sub

    Private Sub SetCurrentValuesForParameterField(ByVal myParameterFields As ParameterFields, ByVal myArrayList As ArrayList)
        Dim currentParameterValues As ParameterValues = New ParameterValues()
        For Each submittedValue As Object In myArrayList
            Dim myParameterDiscreteValue As ParameterDiscreteValue = New ParameterDiscreteValue()
            myParameterDiscreteValue.Value = submittedValue.ToString()
            currentParameterValues.Add(myParameterDiscreteValue)
        Next
        Dim myParameterField As ParameterField = myParameterFields(PARAMETER_FIELD_NAME)
        myParameterField.CurrentValues = currentParameterValues
    End Sub

    Private Sub redisplay_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles redisplay.Click
        Dim myArrayList As ArrayList = New ArrayList()
        For Each item As String In defaultParameterValuesList.SelectedItems
            myArrayList.Add(item)
        Next

        Dim startDate As String = orderStartDate.Text
        Dim endDate As String = orderEndDate.Text

        myCrystalReportViewer.ReportSource = Application.StartupPath & "\" & "CustomersByCity.rpt"
        Dim myParameterFields As ParameterFields = myCrystalReportViewer.ParameterFieldInfo
        SetCurrentValuesForParameterField(myParameterFields, myArrayList)
        SetDateRangeForOrders(myParameterFields, startDate, endDate)
    End Sub

    Private Sub SetDateRangeForOrders(ByVal myParameterFields As ParameterFields, ByVal startDate As String, ByVal endDate As String)
        Dim myParameterRangeValue As ParameterRangeValue = New ParameterRangeValue()
        myParameterRangeValue.StartValue = startDate
        myParameterRangeValue.EndValue = endDate
        myParameterRangeValue.LowerBoundType = RangeBoundType.BoundInclusive
        myParameterRangeValue.UpperBoundType = RangeBoundType.BoundInclusive
        Dim myParameterField As ParameterField = myParameterFields(SUBREPORT_PARAMETER_FIELD_NAME, SUBREPORT_NAME)
        myParameterField.CurrentValues.Clear()
        myParameterField.CurrentValues.Add(myParameterRangeValue)
    End Sub

End Class
